Add a container type to the runtime manager's container status

verb opened the pull request:
**What this PR does / why we need it**:
This is Step 1 of the "Debug Containers" feature proposed in #35584 and is hidden behind a feature gate. Debug containers exist as container status with no associated spec, so this new runtime label allows the kubelet to treat containers differently without relying on spec.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: cc #27140

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
NONE
```

**Integrating feedback**:
- [x] Remove Type field in favor of a help method

**Dependencies:**
- [x] #46261 Feature gate for Debug Containers

verb committed:
Add a container type to the runtime labels

This is part of the "Debug Containers" feature and is hidden behind
a feature gate. Debug containers have no stored spec, so this new
runtime label allows the kubelet to treat containers differently
without relying on spec.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45442)
<!-- Reviewable:end -->

spiffxp commented:
@k8s-bot ok to test

verb commented:
/assign @dashpole

dashpole commented:
Apologies if this has already been hashed out elsewhere, but why do we need a field as part of kubecontainer?  What does the field do that the label does not already do?

For example, the model used for enabling and feature gating "critical pods" uses a [helper method](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/types/pod_update.go#L143) to define a critical pod.  You could define a IsDebugContainer( kubecontainer) helper method, which would avoid adding a field to kubecontainer

verb commented:
@dashpole Nope, no one has brought that up. We don't need the field, it's just that avoiding it didn't occur to me. I like that though and will switch to a helper method.

dashpole commented:
Btw, I don't really have context on the overall proposal, especially some of the design intricacies with how we represent debug containers.  I plan to understand the proposal for the sake of guiding the implementation, but not to work through all of the edge cases associated with it.  With that being said, I'll leave approving the proposal to vish and yuju.  It goes without saying that this PR cannot go in until the proposal is approved.

verb commented:
@dashpole it turns out the kubelet doesn't keep the runtime labels after populating kubecontainer.ContainerStatus, and neither Container nor ContainerStatus have sufficient information to calculate a container type.

Without type being part of ContainerStatus we'd have to query the labels from the runtime, bypassing the cache. It would need to happen from SyncPod(), so it's probably better that we cache it as part of kubecontainer.ContainerStatus.

dashpole commented:
Interesting.  Why use labels if they arent kept here?  I was under the impression that the reason to add this as a label was to pass the container type.

dashpole commented:
It might be nice to schedule a VC just so that you can walk me through how you plan to implement this.

verb commented:
I will definitely do that. In the mean time, there's no reason we actually need to add a field to ContainerStatus in *this* PR where it would be unused. I've moved adding the field to my second PR, which will give context for its use.

I've also updated the unit tests to check for backwards compatibility with containers created prior to the feature being enabled.

dashpole commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -56,6 +60,7 @@ type annotatedPodSandboxInfo struct {
 
 type labeledContainerInfo struct {
 	ContainerName string
+	ContainerType string
```
can we create a separate ContainerType type here?
`type ContainerType string`
`const ContainerType containerTypeInit = "INIT"`
...

verb replied:
yup, Done!

Since the next PR will add a `Type` field to kubecontainer.ContainerStatus which should also be typed, I went ahead and moved the new ContainerType type and constants from kuberuntime to kubecontainer.

dashpole reviewed (commented):
Just one nit.  Otherwise looks good.

dashpole commented:
squash, and then this lgtm

verb commented:
squashed

dashpole commented:
/lgtm

verb commented:
/unassign @vishh @dchen1107 
/assign @yujuhong 

@dashpole @yujuhong This is the first PR of debug containers that David first reviewed back in May, rebased a few times. Since kubernetes/community#649 is merged I think we can pick it up again.

dashpole commented:
Took another look over.
/lgtm

yujuhong commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -94,12 +97,15 @@ func newPodAnnotations(pod *v1.Pod) map[string]string {
 
 // newContainerLabels creates container labels from v1.Container and v1.Pod.
-func newContainerLabels(container *v1.Container, pod *v1.Pod) map[string]string {
+func newContainerLabels(container *v1.Container, pod *v1.Pod, containerType kubecontainer.ContainerType) map[string]string {
```
Do you plan to use the container type as a filter to select containers?
If not, you could also consider writing it as an annotation.

verb replied:
My plan was to add a ContainerType field to kubecontainer.ContainerStatus since most of the places where I need to reference the type I already have the kubecontainer.PodStatus with its list of ContainerStatuses.

I won't need to filter by type unless you think we should avoid the new field in kubecontainer.ContainerStatus. I chose a label of annotation because it "felt" like a more "label-y" thing that someone might want to filter, but it was an arbitrary choice and I'm happy to switch to annotation if you think that's a better fit.

yujuhong commented:
/lgtm
